### DISCUSSION TITLE
Velocitas Create Command Introduction

### DISCUSSION THREAD

**Discussion by kse3hi** [eclipse-velocitas/velocitas-docs#109] [Ideas]
Please leave your impression, suggestion, thoughts as for https://github.com/eclipse-velocitas/velocitas-docs/wiki/Velocitas-create-concept

**Comment by dennismeister93** [comment]
If we want to stay with

```
$ velocitas create \
    --name=VApp \
    --language=python \
    --package=runtimes[@version] \
    --package=devcontainer[@version] \
    --
```

please change ARGS to FLAGS ->

```
USAGE
  $ velocitas create [ARGS...]

ARGS
```

example:

command with args:
`velocitas create vapp python ...`

command with flags:
`velocitas create -n vapp -l python ...`

**Reply by BjoernAtBosch** [reply]
@dennismeister93: Is that Java/Typescript speech?

In general I would talk about arguments and there can be 
* positional arguments, like `command arg1 arg2 ...`,
* options with argument, like `command --file filename`, and
* options w/o argument also called "flags", like `command --help`.

**Reply by doosuu** [reply]
I am also a bit on the fence about naming but Dennis is referring to the OCLIF terms (the CLI framework we are using in the CLI) which refers to options with arguments as `flags` which might be a bit confusing for everyone not familiar with the framework: (see https://oclif.io/docs/flags/)

**Reply by dennismeister93** [reply]
Yes sorry, forgot to mention that my comment relates to the oclif wording, i didn't want to challenge the general wording of args and flags! Thanks @doosuu !

**Reply by kse3hi** [reply]
ARGS changed on FLAGS

**Comment by dennismeister93** [comment]
Is there a specific reason/advantage to use `config.ini` over `config.json`

My thoughts:
1. Almost every config file we introduced is json base (manifest, runtime, velocitas, appmanifest,...)
2. JSON parsing would not introduce an additional "layer" to parse, read and work with .ini file

**Reply by doosuu** [reply]
I am also not a big fan of INI, concept should just describe an abstract configuration structure unless you have a specific reason to chose one format over the other. Since the CLI is a TypeScript application, `JSON` is the better option to use.

**Reply by kse3hi** [reply]
Done

**Comment by dennismeister93** [comment]
It would be good if this concept has a more detailed view on how the actual e2e flow is looking, especially how the folder structure and content of the `vehicle-app-sdk-` should look like when introducing this command.

If i remember correctly the idea was to use these sdk repos as a base for creating the whole VAPP repository and there will be no template anymore, which just gets generated by a CI pipeline?

**Reply by kse3hi** [reply]
We have a requirement section 6.1 that must present this structure.

**Reply by doosuu** [reply]
I'm guessing Dennis is referring to a flow chart which I am also missing.

**Comment by MP91** [comment]
Regarding the requirements I would, for the first step, just focus on the devcontainer, means only Ubuntu **22.04**.

What I completely miss is how we gonna call the command. Do we create a repo just with the devcontainer and the cli installed?

**Reply by kse3hi** [reply]
Corrected

**Reply by MP91** [reply]
You just corrected the Ubuntu version, what about the other things I mentioned?

**Reply by kse3hi** [reply]
Basically, we will use create command to generate `vehicle-app-<lang>-template` therefore it shall be run on devcontainer (as local setup) and on runner as section 5 stays.

**Reply by MP91** [reply]
Please add your proposal for the creation flow. I don't really get it. Do we execute the command locally, in a devcontainer or in a GitHub Workflow?

**Reply by dennismeister93** [reply]
I think 5.2 and 5.3 means that it should be possible to get the velocitas cli (via npm install -g?) locally and execute the create command to generate a new repo? Similar to https://docs.npmjs.com/cli/v9/commands/npm-init.
But if we do it like this. Do we say that "ONLY" the create command works locally and all the others can't be executed? we need to define this

**Reply by MP91** [reply]
Yes but I would not add this _now_ since it means additional effort to make the cli work on the platforms and also additional maintenance effort.

I could imagine a template which just have the cli and the basic devcontainer. When you start the devcontainer it directly starts the creation and than the user does the first commit with all the project stuff

**Reply by doosuu** [reply]
Its better to add a `scope` or `milestone` column which indicates **when** each requirement needs to be implemented. Initially we do not need to run CLI in non-devContainer environments.

**Reply by BjoernAtBosch** [reply]
Couldn't we put the CLI in a container with cli interface? Means, you run it like this:
`docker run velocitas-cli [args]`
where `args` are the arguments/flags to the velocitas cli?

**Reply by MP91** [reply]
I like this idea! We need Docker anyhow and the CLI would work on every OS.

**Reply by doosuu** [reply]
That's more of a concrete technical realization than an abstract requirement. I assume the CLI will just work™ on any OS with the appropriate nodejs version. Wouldn't hurt to release a containerized CLI though.

**Reply by BjoernAtBosch** [reply]
@doosuu: But wouldn't that mean, that one needs a working nodejs installation on their machine? 
I wouldn't like that so much. Docker (or a "compatible" runtime) is anyway a prerequisite and I would prefer not to also add nodejs to our list of prerequisites.

**Reply by MP91** [reply]
It isn't that much of an implementation detail if we just write the requirement like "CLI must work in an Ubuntu based container". Like it is now, it reads as if the cli has to work standalone on Mac and any Linux distribution...

**Comment by doosuu** [comment]
`GITHUB_API_TOKEN` should not be stored in a configuration file but should always be read from the environment variable. We do not want to be responsible to expose secrets if we have this entry as part of our **storeable** configuration.

**Reply by kse3hi** [reply]
Corrected

**Comment by BjoernAtBosch** [comment]
In the "Motivation" section it is stated: "To create a backbone project and automate template generation on GitHub command velocitas create is going to be introduced."

1. What is a backbone project?
2. I thought, we introduce "create" to replace the template instantiation approach by this to create a user project.

At least we should add comma here for better readability:
"To create a backbone project and automate template generation on GitHub**, the** command **`**velocitas create**`** is going to be introduced."

**Reply by kse3hi** [reply]
Applied

**Comment by BjoernAtBosch** [comment]
Requirement 5.2: 
* Why is it just a recommendation and not a must?
* I wouldn't write "should be *available locally*', because that sounds like it should be "there" in every Mac/Linux installation. Better:
`velocitas create` command **shall** be **executable natively at least** on Mac and Linux

**Reply by kse3hi** [reply]
Please see RFC for the meaning of the wording. "should" means "recommended but not mandatory". Can exist can not :)

**Reply by BjoernAtBosch** [reply]
Yes, got that. But my doubt was that it is only a recommendation and not a must.

**Comment by BjoernAtBosch** [comment]
Requirement 5.3:
I think, the proxy "thingy" isn't just related to Windows but at least also to Linux. Don't know if there are BCN Mac clients. (but other companies could have such(?)

**Reply by MP91** [reply]
yes there are also BCN Macs, thats another reason I would only say for now that it should work in a devcontainer, were we already have working settings regarding the proxy stuff...

**Reply by kse3hi** [reply]
Please see RFC, "may" means totally "optional"

**Reply by BjoernAtBosch** [reply]
Yes, maybe I was a bit unspecific. I was thinking if it would be worth to have the need that it should run with and w/o corporate proxy as a separate general requirement. If not we should add this to requirement 5.2 also (because it does not only relate to Windows).

**Comment by BjoernAtBosch** [comment]
Requirement 6.1:
Not sure about its scope or what exactly is meant here.
Requiement 6.1.1:
Isn't that too detailed as a requirement? I would write it in a way that the purpose is clear. (Question: Why can't `create` take it from the vscode folder?)

**Reply by kse3hi** [reply]
Intention to define a structure of the example folder to be taken "as-is".

**Comment by BjoernAtBosch** [comment]
Section "Interactive Mode":
If `velocitas create` is invoked without any arguments, the interactive mode is used.

**Comment by BjoernAtBosch** [comment]
Section "Default Mode":
Isn't the default mode the interactive mode (which is invoked if no arguments)?

**Reply by kse3hi** [reply]
Fixed

**Comment by doosuu** [comment]
What is the meaning of requirement #4?
`To simplify UX-experience default create behavior may be introduced.`

What is meant by default create behavior in this context? Isn't the interactive mode the default behavior?

**Reply by kse3hi** [reply]
Re-phrased

**Comment by doosuu** [comment]
General question:

What about moving away from copying examples from the SDKs and creating a very bare-bones Hello-World type Application? If `vehicle-signal-interface` is not provided by any package, all existing example applications will not work.

**Reply by kse3hi** [reply]
Creation still possible without packages, but has no sense at all.
Additional issue: where to keep and maintain Hello World, in velocitas repo itself?

@dennismeister93 , @MP91 , @BjoernAtBosch  your thoughts?

**Reply by kse3hi** [reply]
If --example is not provided than bare-bone application must be created.
Alt. 1 - place as hidden/normal example (e.g. "Empty")
Alt. 2 - keep it velocitas/elsewhere

**Comment by doosuu** [comment]
`6.1.1 example folder must contain .vscode/launch.json to be reused in creation setup ` 

Can we rephrase this to be a bit more abstract? I.e. `An example folder must be loadable by VSCode without manual steps`? B/c there is more to it than just coping the launch.json

**Comment by MP91** [comment]
Just clarified with Marcel Kurzmann that we need to keep the license file. Since all our files are licensed under Apache it does not make sense to have a different license for the repo itself.
